src/cli/self_update.rs

```diff
@@ -746,21 +748,29 @@ fn maybe_install_rust(
         info!("default host triple is {}", cfg.get_default_host_triple()?);
     }
 
-    // If there is already an install, then `toolchain_str` may not be
-    // a toolchain the user actually wants. Don't do anything.  FIXME:
-    // This logic should be part of InstallOpts so that it isn't
-    // possible to select a toolchain then have it not be installed.
-    if toolchain_str == "none" {
+    let user_specified_something =
+        toolchain.is_some() || !targets.is_empty() || !components.is_empty();
+
+    // If the user specified they want no toolchain, we skip this, otherwise
+    // if they specify something directly, or we have no default, then we install
+    // a toolchain (updating if it's already present) and then if neither of
+    // those are true, we have a user who doesn't mind, and already has an
+    // install, so we leave their setup alone.
```

Comment:
I don't think this is quite right. What if they ask for no tool chain and a component? Should that silently do nothing or error?

Reply:
Well, it won't be silent, but it won't process the invalid input either.

Reply:
other than this, the PR looks good

Reply:
I think that's a good point.  I'd say it's part of `setup_mode.rs` to verify that though.  I've made a note to try and write a test into there.

Reply:
I take that back, clearly it belongs after the user might have made a change.  If they specify targets/components on the CLI and then set the toolchain to none in the interactive prompts, then it still ought to warn.

Reply:
I've pushed a small change, how does that feel to you?